### PR TITLE
Add a checkProbeMergeStatus box just to probe Stash REST API…

### PR DESCRIPTION
**Pull request by jimklimov**
…in order to recalculate refspecs and so work around the Stash feature of lazy reference recalculation as branches evolve. Defaults to "on" because another feature (check for conflicts) that calls same REST API - but can actively refuse to build the PR - defaults to "on".

Hopefully solves https://issues.jenkins-ci.org/browse/JENKINS-35219

### CODE REVIEW

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java
Comment (christ66) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java]: 👎 as this breaks backwards compatibility. Instead it would be best to create a new constructor with the newly added parameter.

**Comment by christ66** [comment]
Other than the new constructor that needs to be made it lgtm.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java
Comment (jimklimov) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java]: Thanks for the review, posting an update, hope I got it right (not a Java programmer). I wonder if it is possible for one constructor to call another (reduce duplication), or is that feature only reserved for normal functions?

I did not realize plugins that implement arbitrary features have some constructor others could depend on, at all :)

**Comment by jimklimov** [comment]
Something breaks about 

````
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project stash-pullrequest-builder: Compilation failure
[ERROR] javax.annotation.processing.FilerException: Attempt to reopen a file for path stash-pullrequest-builder-plugin/target/classes/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.stapler
````

at least for my local `mvn test`... seems the CI here is on-demand somehow?

**Comment by jimklimov** [comment]
Thanks to @daniel-beck for IRC suggestions about passing at least the local `mvn test` - update posted above.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java
Comment (jimklimov) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger.java]: Re-fixed differently after some more suggestions on IRC.

**Comment by jimklimov** [comment]
Pinging @nemccarthy as the only maintainer, to get this PR moving forward :)

**Comment by jakub-bochenski** [comment]
@jimklimov  please rebase the branch against master to get builds or close if you no longer want to work on this.

**Comment by jimklimov** [comment]
I'll take a look, we use this :)

**Comment by jimklimov** [comment]
Rebased over current master.

**Comment by jakub-bochenski** [comment]
Thanks, I'll have to read up on the linked Atlassian answer as I don't understand what the issue is here. 

We do have a PR, approvals and Jenkins builds for them and I haven't noticed any issues with builds not being triggered after a new commit is pushed to the PR branch

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: +1 to comments explaining "why" with links to details

**Comment by proski** [comment]
The code look reasonable, but I would try to avoid adding a new configuration option that is only intended to workaround a known issue. If the option is added, the default should be selected carefully.

I would try to make the plugin to always do the right thing without asking the user. Is there any reasonable situation when the merge status probe would break the functionality? How likely is it?

This is the most intrusive of the open PRs, maybe it should not be part of the release if there is another way to deal with the race condition. Race conditions are hard to test.

Can we recommend some other solution for now? Maybe we could ask users to add a build step to probe the merge status through the Stash REST API? That way, they would be able to test this workaround without installing an unofficial plugin build.

**Comment by jakub-bochenski** [comment]
> This is the most intrusive of the open PRs, maybe it should not be part of the release if there is another way to deal with the race condition.

I plan to do a release with just the fixes merged now, plus the descriptor cleanup. This can go out in the next one.

> Can we recommend some other solution for now? Maybe we could ask users to add a build step to probe the merge status through the Stash REST API? That way, they would be able to test this workaround without installing an unofficial plugin build.

It's possible, but too complicated IMO (requires pre-scm step plugin etc.).
BTW incremental releases/builds are an official channel.

**Comment by jakub-bochenski** [comment]
> Thanks, I'll have to read up on the linked Atlassian answer as I don't understand what the issue is here.

Actually we are observing this issue on projects that doesn't have the "Build only if PR is mergeable" option tuned on.

PS. @jimklimov What I don't get is that "Build only if Stash reports no conflicts" option being turned on does **not** seem to help with this

File: README.md
Comment (jakub-bochenski) [inline README.md]: Can we describe the exact conditions when this workaround is required (AFAIK Build only if PR is mergeable = false)

Comment (jakub-bochenski) [reply README.md]: Please don't paste URLs into the README directly, use the text like 
```
[JENKINS-35219](https://issues.jenkins-ci.org/browse/JENKINS-35219)
```

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (jakub-bochenski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: > Workaround for broken git references that appear due to PRs, popping up in other jobs trying to use just the master branch.

This description is confusing to me, I still don't really understand what is it saying. Maybe something like "Trigger a refresh, because Stash updates the refs lazily"?

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jakub-bochenski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Personally I'm using job-dsl anyway, so it's not affecting me.

Still I think we should make the default false just to keep the previous behavior for sake of least-astonishment.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (proski) [inline src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: The title is very long compared to other titles. It would be better to shorten the name and move the explanation to the help file. This plugin only has help for two fields, see files `help-buildFailedComment.html` and `help-buildFailedComment.html`. The way to add help is described at https://wiki.jenkins.io/display/JENKINS/Basic+guide+to+Jelly+usage+in+Jenkins#BasicguidetoJellyusageinJenkins-Helpfiles
Suggested title: `Probe Stash for merge status`
The help should probably have details about the race condition,

Comment (proski) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: config-screenshot

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (proski) [inline src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: This `if` block of code seems unnecessary and incorrect to me. It the user requests specific checks, they should be performed, and their results should not be ignored. If this block is removed, the code would work correctly. There would be a call to `getPullRequestMergeStatus` (the whole point of this PR), the relevant checks ("can be merged" and "not conflicted") would be performed if enabled, the PR would be considered mergeable if the enabled checks pass or no checks are enabled.

**Comment by proski** [comment]
> I plan to do a release with just the fixes merged now, plus the descriptor cleanup. This can go out in the next one.

Perhaps we should suggest enabling "Build only if Stash reports no conflicts" in README.md. I believe it should also make Stash recalculate the refs. Besides, why would anyone want to build a PR in the conflict state?

**Comment by jimklimov** [comment]
Hello, i hope to get to addressing the comments today or soon after, so
keep them coming :)

Indeed, I believe building merged code should check the mergeability and
recalculate the refspecs.

However people can have valid reasons to build branches that are not
currently mergeable, such as doing experiments and wanting to see how their
verbatim code ends up, whether it builds and logic works in tests, while
they might not be able (skill, security, time, ...) to set up their own
Jenkins job to build their own fork and branch. So they code and open a PR
and maybe the target branch later moves on (perhaps, a competing change is
merged for a quick win but has its downsides so would be replaced by the
solution that takes longer to develop) but they would take care of
conflicts later, or scrap the experiment at all.

At least, that's the sort of usecases that come to my mind as I've seen
them in practice.

Jim

**Comment by jakub-bochenski** [comment]
> Perhaps we should suggest enabling "Build only if Stash reports no conflicts" in README.md. I believe it should also make Stash recalculate the refs.

I re-read the linked atlassian answer, and I think maybe it's possible to make it work without any code changes.
I think the ref that is being updated lazily is only `refs/pull-requests/*/merge`. 

> If you're building refs/pull-requests/*/from, you may choose to notify Jenkins even if isConflicted() returns true; that ref will always exist as long as the pull request is open, even if it has conflicts

If you want to still build the PR even if it has conflicts you should be able to use the `../from` ref. Using the `../merge` ref only makes sense if the PR merges cleanly.

I might have misunderstood something though, @jimklimov ?

**Comment by proski** [comment]
I've tested what would happen if a PR is in the conflicted state. It turns out the their is no `merge` refspec, only `from`. So a job using the `merge` refspec fails, quite expectedly. We should just document that using the `merge` refspec requires "Build only if Stash reports no conflicts". That should take care of the race condition for that case without any code changes.

If the `from` refspec is not affected by the Stash API call we use to determine the PR status, then this PR is pointless. We should fix the documentation instead.

An extra option would only be justified if the `from` refspec is being recalculated by Stash as a result of the API call.

There is still some risk that the hash reported Stash is not the same as the hash fetched by git. I put this on top of all my PR builders

```
fetchedCommitHash="$(git rev-parse refs/remotes/origin/pr/${pullRequestId}/from^{commit})"
if test "${sourceCommitHash}" != "${fetchedCommitHash}"; then
  echo "Stash asked to test commit ${sourceCommitHash}, git fetched commit ${fetchedCommitHash}"
  echo "Refusing to continue to ensure correctness, please retry"
  exit 1
fi
```

It should  be possible to make this plugin check for that mismatch, but that would be a separate effort.

**Comment by jakub-bochenski** [comment]
> An extra option would only be justified if the from refspec is being recalculated by Stash as a result of the API call.

But is it? I think the linked answer suggest it's not affected. 

> There is still some risk that the hash reported Stash is not the same as the hash fetched by git. I put this on top of all my PR builders

As mentioned above I think I've also observed this (on job that checks if the branch is conflicted, but not if it's mergeable). 
If the `../from` ref is buggy, why not just use the `$sourceBranch` instead?

> It should be possible to make this plugin check for that mismatch, but that would be a separate effort.

I think a build trigger is able to do it, you would need to add a build step for that etc.

**Comment by proski** [comment]
> If the `../from` ref is buggy, why not just use the `$sourceBranch` instead?

I'm not implying that any refspec is buggy.

I've seen in many times that `sourceCommitHash` and the `from` refspec are not the same. It happens about once a week to my team. Often enough to be an annoyance, but rare enough to be hard to reproduce and debug.

I was assuming that this PR is dealing with that issue, but now I'm quite sure it's a different issue. We are using the `from` refspec and local merging. We are also using mostly multi-configuration projects, they see most activity and all of the commit mismatches. Hard to draw definite conclusions.

Circling back to selecting what to test. This plugin is responsible for testing a specific revision, `sourceCommitHash`. Stash notifier is configured to report that revision. We cannot just test one revision and report another, that's defeating the whole purpose of continuous integration.

The only acceptable behavior is testing `sourceCommitHash` or its immediate merge child (the later only if expected and allowed). I don't want any guesses. Ensuring that would be a separate effort.

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (jimklimov) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: tried to address this now

File: README.md
Comment (jimklimov) [reply README.md]: tried to address this now

Comment (jimklimov) [reply README.md]: Not so sure about clarifying the exact conditions, though... something was buggy, I investigated and fixed that, but got no certain science for that. :-)

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jimklimov) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Not sure... it is an intrusive operation a bit (causing load on Stash), but if you poll this repo from a job you also want to be certain to catch new commits using up to date refspecs and in general look at a consistent data view.

File: README.md
Comment (jakub-bochenski) [reply README.md]: OK, please check the discussion here too https://github.com/jenkinsci/stash-pullrequest-builder-plugin/pull/22#issuecomment-460199716
Based on the research @proski  did and on the atlassian answer I'm suspecting there might be a problem with the build using stale commit, but it would *not* be fixed by triggering via the REST API

File: src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java
Comment (jimklimov) [reply src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java]: Thanks, your point does sound valid, especially if those other checks were requested and did fail.

I'll convert the `if` to just a big comment about why that toggle is consulted in the outer big `if` but not inside the `res` evaluation.

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jimklimov) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Thanks, I thought of that too, as I got acquainted with Jelly+Help usage by example in another plugin :)

File: README.md
Comment (jimklimov) [reply README.md]: I cannot really constructively comment on that... my guess would be that there are many different versions (and separate codebases) out there sharing the Bitbucket/Stash name and history, so behaviors probably vary. With the one(s) we have, there were problems using a wrong commit before this change (as posted originally), and none noted in the year and a half since then.

So my best guess is that at least for the deployment we have, it does solve the issue :-)

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jimklimov) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: Posted the commit.

**Comment by jimklimov** [comment]
Thanks @jakub-bochenski and @proski for the reviews, I hope I addressed your concerns the best I could.

File: README.md
Comment (jakub-bochenski) [inline README.md]: minor: I think this can be shortened, the first two sentences kind of say the same, why not just `Probe the Stash REST API for merge status just so it recalculates the refspec: see [JENKINS-35219](https://issues.jenkins-ci.org/browse/JENKINS-35219) ... `

Comment (jimklimov) [reply README.md]: Good catch. The part before the colon is the (later shortened) option name as seen in configs. The rest is layman's description (a revision of which is now also the help text for the option).

File: README.md
Comment (jakub-bochenski) [reply README.md]: That's good for you :), but a good description will also help other people.
Say somebody may have the same problem, but not realize it's a workaround for it.

If I guess your case correctly, maybe change `Has no impact on proceeding with the build itself.` to `Use this if you encounter problems with stale commits being built, but don't want to skip builds based on the PR status (like the two options above)`

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jakub-bochenski) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: As discussed in the other thread this might not be generally useful, only for some specific configurations or stash versions. That would be another argument for making it disabled by default

File: README.md
Comment (jakub-bochenski) [reply README.md]: Also it might be useful to clarify this has no effect if either of the two options above are enabled.

Comment (jimklimov) [reply README.md]: done, hopefully

File: src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly
Comment (jimklimov) [reply src/main/resources/stashpullrequestbuilder/stashpullrequestbuilder/StashBuildTrigger/config.jelly]: OK, you convinced me :)

**Comment by jimklimov** [comment]
Huh, resolving merge conflicts pulled in the current master... oh well, can eventually "Rebase and merge" this one, I guess.

**Comment by proski** [comment]
Thank you for the cleanup, it looks much better now. Special thanks for the help text. If would be nice to provide help to all options (as a separate PR, of course). I'll try to get this compiled and tested on my Jenkins installation shortly.

In the meantime, I'm using Branch Specifier `${sourceCommitHash}` and `Build only if Stash reports no conflicts`. It's been a few days since I implemented the change, and my team is happy with it. Nobody is surprised why conflicted PRs are not being tested.

**Comment by jimklimov** [comment]
Well, four eyes see better than two, six even more so :)

Thank you both for feedback and nudges, hope this soon gets into an
agreeable shape to get merged too :)

**Review by jakub-bochenski — APPROVED** [review]

**Comment by proski** [comment]
Any chances to have a release now?

It would be nice to have help for all options, but I don't see as a release blocker.

Please make sure not to use version 1.7.1 to avoid confusion, it looks like it was "released" and then "unreleased". Considering the change of the maintainers, version 1.8.0 would be justified.

**Pull request merged**
